wen-restart/src/wen_restart.rs

```diff
@@ -405,6 +415,151 @@ fn find_bankhash_of_heaviest_fork(
+    if progress.my_heaviest_fork.is_none() {
+        return Err(WenRestartError::UnexpectedState(RestartState::HeaviestFork).into());
+    }
+    let my_heaviest_fork = progress.my_heaviest_fork.clone().unwrap();
+    let heaviest_fork_slot = my_heaviest_fork.slot;
+    let heaviest_fork_hash = Hash::from_str(&my_heaviest_fork.bankhash)?;
+    let mut heaviest_fork_aggregate = HeaviestForkAggregate::new(
+        wait_for_supermajority_threshold_percent,
+        cluster_info.my_shred_version(),
+        epoch_stakes,
+        heaviest_fork_slot,
+        heaviest_fork_hash,
+        &cluster_info.id(),
+    );
+    if let Some(aggregate_record) = &progress.heaviest_fork_aggregate {
+        for (key_string, message) in &aggregate_record.received {
+            if let Err(e) = heaviest_fork_aggregate.aggregate_from_record(key_string, message) {
+                error!("Failed to aggregate from record: {:?}", e);
+            }
+        }
+    } else {
+        progress.heaviest_fork_aggregate = Some(HeaviestForkAggregateRecord {
+            received: HashMap::new(),
+            final_result: None,
+        });
+    }
+
+    let mut total_active_stake = heaviest_fork_aggregate.total_active_stake();
+    progress
+        .my_heaviest_fork
+        .as_mut()
+        .unwrap()
+        .total_active_stake = total_active_stake;
+    cluster_info.push_restart_heaviest_fork(
+        heaviest_fork_slot,
+        heaviest_fork_hash,
+        total_active_stake,
+    );
+
+    let mut cursor = solana_gossip::crds::Cursor::default();
+    loop {
+        if exit.load(Ordering::Relaxed) {
+            return Err(WenRestartError::Exiting.into());
+        }
+        let mut progress_changed = false;
+        let start = timestamp();
+        for new_heaviest_fork in cluster_info.get_restart_heaviest_fork(&mut cursor) {
+            let from = new_heaviest_fork.from.to_string();
+            if let Some(record) = heaviest_fork_aggregate.aggregate(new_heaviest_fork) {
+                progress
+                    .heaviest_fork_aggregate
+                    .as_mut()
+                    .unwrap()
+                    .received
+                    .insert(from, record);
+                progress_changed = true;
+            }
+        }
+        let current_total_active_stake = heaviest_fork_aggregate.total_active_stake();
+        if current_total_active_stake > total_active_stake {
+            total_active_stake = current_total_active_stake;
+            cluster_info.push_restart_heaviest_fork(
+                heaviest_fork_slot,
+                heaviest_fork_hash,
+                current_total_active_stake,
+            );
+            progress
+                .my_heaviest_fork
+                .as_mut()
+                .unwrap()
+                .total_active_stake = current_total_active_stake;
+            progress_changed = true;
+        }
+        if progress_changed {
+            write_wen_restart_records(wen_restart_path, progress)?;
```

Comment:
This seems unnecessarily expensive to write the entire record every time instead of just the diff.

Reply:
this content in the file is a snapshot of our progress, it is a bit expensive, but if we change to Gossip changes every 30 minutes maybe the cost is okay?

Reply:
isn't it written on every update from every potential validator? So potentially could be hundreds of rewrites in the worst case within a 30 minute period

Reply:
Nevermind, I see, it was wrapped around in the 30 minute update change

Reply:
Although this avoids a bunch of writes, my gut feeling is it would be useful to log or write out every individual validator's heaviest fork for debugging purposes in case we ever get stuck

Reply:
Added info log.